sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -1198,8 +1209,34 @@ def process(self, element, *schema_side_inputs):
       return self._flush_all_batches()
 
   def finish_bundle(self):
+    current_millis = int(time.time() * 1000)
+    if BigQueryWriteFn.LATENCY_LOGGING_LOCK.acquire(False):
```

Comment:
Help me understand a bit here. Am I correct about this:
- There are multiple threads
- But due to Global Interpreter Lock, only one thread runs at a time
- But threads can possibly block in these sections, get preempted and another thread will start running
- so we need the locking to ensure only one thread is running in this block at a time.

In addition
- Multiple python processes also execute, and there are multiple python SDK harnesses running in order to get more parallelism (they don't have the same type of threading concurrency issues, but may communicate with the multiprocessing module).
- Though this is not why the lock is introduced

Reply:
Lock was introduced for sharing the same histogram object between threads in a single python process. My first attempt was having a separate histogram object per DoFn but it produced too many logs since we're creating dozens of DoFns per process with streaming pipelines (# of workers * # of cores * # of DoFn threads).

There are two locks: the first one is for synchronizing histogram object itself and the second one is only for printing out the logs with a given interval. For the second lock, other threads don't need to wait if one thread occupies the lock. It's okay that any one of threads puts its hands up first and does the job and others just skip the whole code block.